core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.storage
+
+import scala.collection.mutable
+
+import org.apache.spark.scheduler._
+
+private[spark] case class BlockUIData(
+    blockId: BlockId,
+    storageLevel: StorageLevel,
+    memSize: Long,
+    diskSize: Long,
+    externalBlockStoreSize: Long,
+    locations: Set[String])
+
+private[spark] class BlockStatusListener extends SparkListener {
+
+  private val blockManagers = new mutable.HashMap[BlockManagerId, mutable.HashSet[BlockId]]
+  private val blocks = new mutable.HashMap[BlockId, BlockUIData]
+
+  override def onBlockUpdated(blockUpdated: SparkListenerBlockUpdated): Unit = {
+    val blockId = blockUpdated.blockUpdatedInfo.blockId
+    if (!blockId.isInstanceOf[StreamBlockId]) {
+      // Now we only monitor StreamBlocks
+      return
+    }
+    val blockManagerId = blockUpdated.blockUpdatedInfo.blockManagerId
+    val storageLevel = blockUpdated.blockUpdatedInfo.storageLevel
+    val memSize = blockUpdated.blockUpdatedInfo.memSize
+    val diskSize = blockUpdated.blockUpdatedInfo.diskSize
+    val externalBlockStoreSize = blockUpdated.blockUpdatedInfo.externalBlockStoreSize
+
+    synchronized {
+      // Drop the update info if the block manager is not registered
+      blockManagers.get(blockManagerId).foreach { blocksInBlockManager =>
+        if (storageLevel.isValid) {
+          blocksInBlockManager.add(blockId)
+          val location = blockManagerId.hostPort
+          val newLocations =
+            blocks.get(blockId).map(_.locations).getOrElse(Set.empty) + location
+          val newStorageLevel = StorageLevel(
+            useDisk = diskSize > 0,
+            useMemory = memSize > 0,
+            useOffHeap = externalBlockStoreSize > 0,
```

Comment:
Yikes! Yeah we cant break sorting like that. Hmmm .. we need to think from scratch. Unlike RDDs, blocks can only be in one level per location. So showing "size in memory" AND "size in disk" does not make sense. It made sense for RDDs because some partitions may be in memory and some may be in disk. Let me thinks whats a better organization.